common/scala/src/main/scala/org/apache/openwhisk/core/yarn/YARNContainerFactory.scala

```diff
@@ -0,0 +1,476 @@
+    var started = false
+    var retryCount = 0
+    val maxRetryCount = serviceStartTimeoutMS / retryWaitMS
+    while (!started && retryCount < maxRetryCount) {
+      val serviceDef = downloadServiceDefinition()
+
+      //Should never happen
+      if (serviceDef == null)
+        throw new IllegalArgumentException("Service not found?")
+
+      serviceDef.state match {
+        case "STABLE" | "STARTED" =>
+          logging.info(this, "YARN service achieved stable state")
+          started = true
+
+        case state =>
+          logging.info(
+            this,
+            s"YARN service is not in stable state yet ($retryCount/$maxRetryCount). Current state: $state")
+          Thread.sleep(retryWaitMS)
+      }
+      retryCount += 1
+    }
+    if (!started)
+      throw new Exception(s"After ${containerStartTimeoutMS}ms YARN service did not achieve stable state")
+  }
+  def waitForContainerStability(actionImage: ImageName): ServiceDefinition = {
+    var stable = false
+    var serviceDef: ServiceDefinition = null
+    var retryCount = 0
+    val maxRetryCount = containerStartTimeoutMS / retryWaitMS
+    while (!stable && retryCount < maxRetryCount) {
+      serviceDef = downloadServiceDefinition()
+      val newContainerList = serviceDef.components
+        .filter(c => c.name.equals(actionImage.name))
+        .flatMap(c => c.containers.getOrElse(List[ContainerDefinition]()))
+
+      val previousContainerList = containerMap.getOrElse(actionImage.name, List[ContainerDefinition]())
+
+      //If container has not been created yet, or container is not in ready state yet
+      if (newContainerList.size > previousContainerList.size && newContainerList.forall(c => c.state.equals("READY"))) {
+        logging.info(this, s"YARN container (${actionImage.name}) achieved stable state")
+        stable = true
+      } else {
+        logging.info(
+          this,
+          s"YARN container (${actionImage.name}) is not in stable state yet ($retryCount/$maxRetryCount)")
+      }
+      Thread.sleep(retryWaitMS)
+      retryCount += 1
+    }
+    if (stable)
+      serviceDef
+    else
+      throw new Exception(s"After ${containerStartTimeoutMS}ms YARN container did not achieve stable state")
+  }
+
+  def downloadServiceDefinition(): ServiceDefinition = {
+    val response: httpresponse =
+      submitRequestWithAuth(HttpMethods.GET, s"${yarnConfig.masterUrl}/app/v1/services/${yarnConfig.serviceName}", "")
+
+    response match {
+      case httpresponse(StatusCodes.OK, content) =>
+        //logging.info(this,"Retrieved service definition from YARN")
+        content.parseJson.convertTo[ServiceDefinition]
+
+      case httpresponse(StatusCodes.NotFound, content) =>
+        logging.info(this, s"Service not found. Response: $content")
+        null
+
+      case httpresponse(_, _) =>
+        handleYARNRESTError(response)
+        null
+    }
+  }
+
+  def removeContainer(task: YARNTask): Unit = {
+
+    serviceLock.synchronized {
```

Comment:
the docker factory also have a mechanism for synchronizing docker operations (there's a way to limit the number of concurrent operations) - it might be worth looking at how that's implemented for reference.